pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/ComplexTypeTransformer.java

```diff
@@ -0,0 +1,184 @@
+        list.add(record);
+      } else {
+        for (Object obj : (Object[]) value) {
+          GenericRow copy = flattenCollectionItem(record, obj, column);
+          list.add(copy);
+        }
+      }
+    }
+  }
+
+  private GenericRow flattenCollectionItem(GenericRow record, Object obj, String column) {
+    GenericRow copy = record.copy();
+    if (obj instanceof Map) {
+      Map<String, Object> map = (Map<String, Object>) obj;
+      for (Map.Entry<String, Object> entry : new HashSet<>(map.entrySet())) {
+        String flattenName = concat(column, entry.getKey());
+        copy.putValue(flattenName, entry.getValue());
+      }
+    } else {
+      copy.putValue(column, obj);
+    }
+    return copy;
+  }
+
+  @VisibleForTesting
+  protected GenericRow flattenMap(GenericRow record, Collection<String> columns) {
+    for (String column : columns) {
+      if (record.getValue(column) instanceof Map) {
+        Map<String, Object> map = (Map) record.removeValue(column);
+        List<String> mapColumns = new ArrayList<>();
+        for (Map.Entry<String, Object> entry : new HashSet<>(map.entrySet())) {
+          String flattenName = concat(column, entry.getKey());
+          record.putValue(flattenName, entry.getValue());
+          mapColumns.add(flattenName);
+        }
+        record = flattenMap(record, mapColumns);
+      } else if (record.getValue(column) instanceof Collection && _collectionsToUnnest.contains(column)) {
+        for (Object inner : (Collection) record.getValue(column)) {
+          if (inner instanceof Map) {
+            Map<String, Object> innerMap = (Map<String, Object>) inner;
+            flattenMap(column, innerMap, new HashSet<>(innerMap.keySet()));
+          }
+        }
+      }
+    }
+    return record;
+  }
+
+  private void flattenMap(String path, Map<String, Object> map, Collection<String> fields) {
+    for (String field : fields) {
+      if (map.get(field) instanceof Map) {
+        Map<String, Object> innerMap = (Map<String, Object>) map.remove(field);
+        List<String> innerMapFields = new ArrayList<>();
+        for (Map.Entry<String, Object> innerEntry : new HashSet<>(innerMap.entrySet())) {
+          map.put(concat(field, innerEntry.getKey()), innerEntry.getValue());
+          innerMapFields.add(concat(field, innerEntry.getKey()));
+        }
+        if (!innerMapFields.isEmpty()) {
+          flattenMap(concat(path, field), map, innerMapFields);
+        }
+      } else if (map.get(field) instanceof Collection && _collectionsToUnnest.contains(concat(path, field))) {
+        for (Object inner : (Collection) map.get(field)) {
+          if (inner instanceof Map) {
+            Map<String, Object> innerMap = (Map<String, Object>) inner;
+            flattenMap(concat(path, field), innerMap, new HashSet<>(innerMap.keySet()));
+          }
+        }
+      }
+    }
+  }
+
+  private String concat(String left, String right) {
+    return String.join(DELIMITER, left, right);
```

Comment:
Make this static. Higher chance of JVM inlining it

Reply:
done